sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/rules/RuleExecutor.scala

```diff
@@ -98,7 +107,8 @@ abstract class RuleExecutor[TreeType <: TreeNode[_]] extends Logging {
         if (iteration > batch.strategy.maxIterations) {
           // Only log if this is a rule that is supposed to run more than once.
           if (iteration != 2) {
-            logInfo(s"Max iterations (${iteration - 1}) reached for batch ${batch.name}")
+            val msg = s"Max iterations (${iteration - 1}) reached for batch ${batch.name}"
+            if (batch.strategy.throws) throw new SparkException(msg) else logTrace(msg)
```

Comment:
This isn't the users fault and will only happen in tests so I wouldn't throw `AnalysisException`.  I might even assert or something?